core/src/main/java/com/linecorp/armeria/common/RequestContextAwareCompletableFuture.java

```diff
@@ -243,4 +243,92 @@
+                                              Executor executor) {
+        return ctx.makeContextAware(supplyAsync(this.makeContextAware(supplier), executor));
+    }
+
+    private Supplier<T> makeContextAware(Supplier<? extends T> action) {
+        return () -> {
+            try (SafeCloseable ignored = ctx.pushIfAbsent()) {
+                return action.get();
+            }
+        };
+    }
+
+    private MinimalStage<T> uniAsMinimalStage() {
+        return new MinimalStage<>();
+    }
+
+    static final class MinimalStage<T> extends CompletableFuture<T> {
+        MinimalStage() {}
+        public <U> CompletableFuture<U> newIncompleteFuture() {
+            return new MinimalStage<>(); }
+        @Override public T get() {
+            throw new UnsupportedOperationException(); }
+        @Override public T get(long timeout, TimeUnit unit) {
+            throw new UnsupportedOperationException(); }
+        @Override public T getNow(T valueIfAbsent) {
+            throw new UnsupportedOperationException(); }
+        @Override public T join() {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean complete(T value) {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean completeExceptionally(Throwable ex) {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean cancel(boolean mayInterruptIfRunning) {
+            throw new UnsupportedOperationException(); }
+        @Override public void obtrudeValue(T value) {
+            throw new UnsupportedOperationException(); }
+        @Override public void obtrudeException(Throwable ex) {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean isDone() {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean isCancelled() {
+            throw new UnsupportedOperationException(); }
+        @Override public boolean isCompletedExceptionally() {
+            throw new UnsupportedOperationException(); }
+        @Override public int getNumberOfDependents() {
+            throw new UnsupportedOperationException(); }
+        public CompletableFuture<T> completeAsync
+                (Supplier<? extends T> supplier, Executor executor) {
+            throw new UnsupportedOperationException(); }
+        public CompletableFuture<T> completeAsync
+                (Supplier<? extends T> supplier) {
+            throw new UnsupportedOperationException(); }
+        public CompletableFuture<T> orTimeout
+                (long timeout, TimeUnit unit) {
+            throw new UnsupportedOperationException(); }
+        public CompletableFuture<T> completeOnTimeout
+                (T value, long timeout, TimeUnit unit) {
+            throw new UnsupportedOperationException(); }
+        @Override public CompletableFuture<T> toCompletableFuture() {
+            if (super.isDone()) {
```

Comment:
I think this this can be context aware by doing

```java
CompletableFuture<T> future = newIncompleteFuture();
whenComplete((value, t) -> {
  if (t != null) {
    future.completeExceptionally(t);
 } else {
   future.complete(value);
  }
  return future;
}
```

I guess us propagating whenComplete into a new future is similar to the JDK's private methods using `uniPush`.